### PR TITLE
Prevent using graphtype as model

### PR DESCRIPTION
**Pull request by Shane32**
Implemented within `ComplexGraphType<TSourceType>`

Applies to:
- `ObjectGraphType<TSourceType>`
- `InputObjectGraphType<TSourceType>`
- `AutoRegisteringObjectGraphType<TSourceType>`
- `AutoRegisteringInputObjectGraphType<TSourceType>`

### CODE REVIEW

**Comment by Shane32** [comment]
This PR doesn't work because `__Type` derives from `ObjectGraphType<IGraphType>`.  Options:

1. Change `__Type` to derive from `ObjectGraphType` rather than `ObjectGraphType<IGraphType>`.  This would technically be a breaking change
2. Move the checks into the auto-registering graph type constructors

**Comment by Shane32** [comment]
Solution 2 implemented here

File: src/GraphQL.Tests/Bugs/BubbleUpTheNullToNextNullable.cs [hunk 1]
```diff
@@ -268,21 +268,26 @@ public BubbleNullSchema()
...
 
         query.Field<NonNullGraphType<ListGraphType<NonNullGraphType<DataGraphType>>>>("nonNullableListOfNonNullableDataGraph")
-            .Resolve(_ => new[] { new DataGraphType() });
+            .Resolve(_ => new[] { new DataModel() });
 
         Query = query;
     }
 }
 
-public class DataGraphType : ObjectGraphType<DataGraphType>
```
Comment (sungam3r) [inline src/GraphQL.Tests/Bugs/BubbleUpTheNullToNextNullable.cs:285]: Ha)

Comment (sungam3r) [reply src/GraphQL.Tests/Bugs/BubbleUpTheNullToNextNullable.cs:285]: That means that in theory in very specific cases someone **may** want to inherit from `ObjectGraphType<OtherGraphType>`.

File: src/GraphQL/Types/Composite/ComplexGraphType.cs [hunk 2]
```diff
@@ -17,6 +17,9 @@ public abstract class ComplexGraphType<TSourceType> : GraphType, IComplexGraphTy
         /// <inheritdoc/>
         protected ComplexGraphType()
         {
+            if (typeof(IGraphType).IsAssignableFrom(typeof(TSourceType)) && GetType() != typeof(Introspection.__Type))
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/ComplexGraphType.cs:20]: 🤔 Maybe in future we introduce some flag (GlobalSwitch?) to allow this behavior.

Comment (sungam3r) [reply src/GraphQL/Types/Composite/ComplexGraphType.cs:20]: I understand that for 99,99% cases this check helps to solve developer problems.

File: src/GraphQL.Tests/Types/NonNullGraphTypeTests.cs [hunk 3]
```diff
@@ -94,34 +94,35 @@ public NullableSchema()
...
+            .Resolve(c => new DataModel { Data = c.Source as ExampleContext });
         query.Field<NonNullableSchemaType>("nonNullable")
-            .Resolve(c => new NonNullableSchemaType { Data = c.Source as ExampleContext });
+            .Resolve(c => new DataModel { Data = c.Source as ExampleContext });
 
         Query = query;
     }
 }
 
-public class NullableSchemaType : ObjectGraphType<NullableSchemaType>
```
Comment (sungam3r) [inline src/GraphQL.Tests/Types/NonNullGraphTypeTests.cs:105]: Ah, even one more place.

**Comment by Shane32** [comment]
Maybe a global switch. I figured that (a) if there’s a problem then someone will raise an issue (b) people can use the non-generic version of ObjectGraphType if necessary in these rare circumstances. 

I suggest we wait until someone complains before adding a global switch.

File: src/GraphQL.Tests/Types/NonNullGraphTypeTests.cs (hunk 3 above)
Comment (Shane32) [reply src/GraphQL.Tests/Types/NonNullGraphTypeTests.cs:105]: These are tests, after all - would they really be coded like this for production use? I highly doubt it — running all the graph constructor code for a data model? Seems highly unlikely. 

Could be though - I’m not sure.

Comment (sungam3r) [reply src/GraphQL.Tests/Types/NonNullGraphTypeTests.cs:105]: Tests, yes. We ourselves write a code that should not be written.

**Comment by sungam3r** [comment]
Of course.

**Pull request merged**